test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -49,43 +57,72 @@ public static void main(String[] args) throws IOException {
+                    DatagramChannel.open().bind(refuser).close();
+                    throw new RuntimeException("Unexpected wakeup");
+                }
+                break;
             }
-
             /* Test 2: connected so ICMP port unreachable may be received */
             dc.connect(refuser);
             try {
-                sendDatagram(dc, refuser);
-                n = sel.select(2000);
-                if (n > 0) {
-                    sel.selectedKeys().clear();
-                    try {
-                        n = dc.read(ByteBuffer.allocate(100));
-                        throw new RuntimeException("Unexpected datagram received");
-                    } catch (PortUnreachableException pue) {
-                        // expected
+                for (int i = 0; i < MAX_TRIES; i++) {
+                    sendDatagram(dc, refuser);
+                    int n = sel.select(2000);
+                    if (n > 0) {
+                        sel.selectedKeys().clear();
+                        try {
+                            ByteBuffer buf = ByteBuffer.allocate(100);
+                            SocketAddress sa = dc.receive(buf);
+                            buf.flip();
+                            byte[] bytes = new byte[buf.remaining()];
+                            buf.get(bytes);
+                            String message = new String(bytes);
+                            System.out.format("received %s at %s from %s%n", message, dc.getLocalAddress(), sa);
+                            //If any received data contains the message from sendDatagram then return true
+                            if (message.contains(GREETINGS_MESSAGE)) {
+                                throw new RuntimeException("Unexpected datagram received");
+                            }
+                            // BindException will be thrown if another service is using
+                            // our expected refuser port, cannot run just exit.
+                            DatagramChannel.open().bind(refuser).close();
+                            if (i < MAX_TRIES - 1) continue;
+                            throw new RuntimeException("PortUnreachableException not raised");
+                        } catch (PortUnreachableException pue) {
+                            System.out.println("Got expected PortUnreachableException " + pue);
+                        }
                     }
+                    break;
                 }
             } finally {
                 dc.disconnect();
             }
-
-            /* Test 3: not connected so ICMP port unreachable should not be received */
-            sendDatagram(dc, refuser);
-            n = sel.select(2000);
-            if (n > 0) {
-                throw new RuntimeException("Unexpected wakeup after disconnect");
+            for (int i = 0; i < MAX_TRIES; i++) {
+                /* Test 3: not connected so ICMP port unreachable should not be received */
```

Comment:
For readability it might be better to restore the break between Test 2 and Test 3 so that it's easier to see where Test 3 starts. The updated test is a bit hard to follow because the original comments have been moved into the loop and maybe better to just leave them where they were, as in Test 2.